[Hazmat Shipping] Add hazmat analytics tracks

ThomazFB opened the pull request:
Summary
==========
Introduce a set of Tracks to the Shipping Label creation flow regarding Hazmat package declarations. The following tracks were added:

- CONTAINS_HAZMAT_CHECKED -> When the `Contains Hazardous Materials` toggle is activated
- HAZMAT_CATEGORY_SELECTOR_OPENED -> When the Hazmat category selector is displayed to the user
- HAZMAT_CATEGORY_SELECTED -> When the user selects an Hazmat category
   **Parameters**: 
   - Category: The selected category
   - Order ID: ID of the order with hazmat content under the Shipping Label creation process

How to Test
==========
1. Open the order details of an order eligible for Shipping label creation
2. Start the shipping label creation flow with valid US address options until you reach the Package details section
3. Inside the package details view, make sure the Contains hazardous materials toggle is visible and interactable
4. Hit the `Contains Hazardous Materials` toggle and verify that `🔵 Tracked: contains_hazmat_checked` is triggered.
5. Make sure the toggle expands and collapses a basic view that will serve as the main control options for the HAZMAT declaration
6. Hit the `Select a category` button and make sure the `🔵 Tracked: hazmat_category_selector_opened` is triggered.
7. Select a hazmat option and make sure the `🔵 Tracked: hazmat_category_selected` is triggered.

ThomazFB committed:
Add analytics dependency to ShippingLabelSinglePackageViewModel

ThomazFB committed:
Introduce hazmat events

ThomazFB committed:
Add hazmat track event triggers to viewModel

ThomazFB committed:
Adjust hazmat selector track event call

ThomazFB committed:
Add hazmat event actions into WooAnalyticsEvent type

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Fix build issues

ThomazFB committed:
Replace all direct stats references with WooAnalyticsEvent declarations

ThomazFB committed:
Fix multiple build issues

ThomazFB committed:
Add hazmat category selection track parameters

ThomazFB committed:
Merge branch 'trunk' into issue/hazmat-tracks

ThomazFB committed:
Fix hound issues

ThomazFB committed:
Add analytics mock configuration to single package tests

ThomazFB committed:
Remove unnecessary weak self declaration

ThomazFB committed:
Introduce initial test implementation for single package analytics

ThomazFB committed:
Add more analytics test to Single Package ViewModel

ThomazFB committed:
Finish unit tests for Single Package analytics

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/Create Shipping Label/Multi-package/ShippingLabelSinglePackageViewModelTests.swift:
```diff
@@ -7,6 +7,15 @@ class ShippingLabelSinglePackageViewModelTests: XCTestCase {
+    private var analyticsProvider: MockAnalyticsProvider!
+    private var analytics: Analytics!
+
+    override func setUp() {
+        super.setUp()
+        analyticsProvider = MockAnalyticsProvider()
+        analytics = WooAnalytics(analyticsProvider: analyticsProvider)
+    }
```
nit: usually we also nil the variables out in the reverse order in `tearDown`, just to clear all the states between tests

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/Create Shipping Label/Multi-package/ShippingLabelSinglePackageViewModelTests.swift:
```diff
@@ -547,6 +556,92 @@ class ShippingLabelSinglePackageViewModelTests: XCTestCase {
         XCTAssertEqual(viewModel.selectedHazmatCategory, .none)
     }
+
+    func test_when_changing_hazmat_category_then_expected_analytics_is_triggered() throws {
```
nice tests!

jaclync commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Multi-package/ShippingLabelSinglePackageViewModel.swift:
```diff
@@ -246,9 +253,20 @@ final class ShippingLabelSinglePackageViewModel: ObservableObject, Identifiable
+            .filter { $0 != .none }
+            .sink { [weak self] selectedCategory in
+                guard let order = self?.order else { return }
+                self?.analytics.track(event: .ShippingLabelHazmatDeclaration.hazmatCategorySelected(orderID: order.orderID,
```
nit: might look nicer if we guard self here as well to avoid more optional self later
```suggestion
                guard let self else { return }
                self.analytics.track(event: .ShippingLabelHazmatDeclaration.hazmatCategorySelected(orderID: self.order.orderID,
```

jaclync reviewed (approved):
LGTM! :shipit: 
It's interesting how I kept getting a 500 error when trying to get carriers & rates unless I selected Hazmat category 1, did you also see this?

```
{
	"error": "wcc_server_error_response",
	"message": "Error: The WooCommerce Shipping & Tax server returned: Bad Request child \"packages\" fails because [\"packages\" at position 0 does not match any of the allowed types] ( 400 )"
}
```

ThomazFB committed:
Update WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Multi-package/ShippingLabelSinglePackageViewModel.swift

Co-authored-by: Jaclyn Chen <watertranquil@gmail.com>

ThomazFB commented:
Once again, thank you so much for your quick and relevant reviews, Jaclyn! 🙇 I submitted your suggestions!

> LGTM! :shipit: It's interesting how I kept getting a 500 error when trying to get carriers & rates unless I selected Hazmat category 1, did you also see this?

Thank you for reporting this, Jaclyn 🙏 . I actually started the fix of this problem right away this morning because Anitaa raised the same flag. I investigated what happened and I noticed that I introduced a regression into the iOS when I changed the Hazmat category parameter handling inside the `ShippingLabelPackageSelected` type. The fix is already under review for release 15.5 in https://github.com/woocommerce/woocommerce-ios/pull/10805

ThomazFB committed:
Add tearDown into Single Package tests class
